AwsCryptographyPrimitives/runtimes/go/ImplementationFromDafny-go/RSAEncryption/externs.go

```diff
@@ -0,0 +1,139 @@
+	derPublicKey, rest := pem.Decode(dafny.ToByteArray(publicKeyInput))
+	if len(rest) > 0 {
+		return Wrappers.Companion_Result_.Create_Failure_(AwsCryptographyPrimitivesTypes.Companion_Error_.Create_AwsCryptographicPrimitivesError_(
+			dafny.SeqOfChars([]dafny.Char(fmt.Errorf("failed to decode PEM: invalid bytes:  %s", rest).Error())...)))
+	}
+	publicKey, err := x509.ParsePKIXPublicKey(derPublicKey.Bytes)
+	if err != nil {
+		return Wrappers.Companion_Result_.Create_Failure_(AwsCryptographyPrimitivesTypes.Companion_Error_.Create_AwsCryptographicPrimitivesError_(
+			dafny.SeqOfChars([]dafny.Char(err.Error())...)))
+	}
+	return Wrappers.Companion_Result_.Create_Success_(uint32(publicKey.(*rsa.PublicKey).N.BitLen()))
+}
+
+func (CompanionStruct_Default___) DecryptExtern(paddingMode AwsCryptographyPrimitivesTypes.RSAPaddingMode,
+		key dafny.Sequence, cipher_text dafny.Sequence) Wrappers.Result {
+	derPrivateKey, rest := pem.Decode(dafny.ToByteArray(key))
+	if len(rest) > 0 {
+		return Wrappers.Companion_Result_.Create_Failure_(AwsCryptographyPrimitivesTypes.Companion_Error_.Create_AwsCryptographicPrimitivesError_(
+			dafny.SeqOfChars([]dafny.Char(fmt.Errorf("failed to decode PEM: invalid bytes:  %s", rest).Error())...)))
+	}
+
+	privateKey, err := x509.ParsePKCS8PrivateKey(derPrivateKey.Bytes)
+	if err != nil {
+		return Wrappers.Companion_Result_.Create_Failure_(AwsCryptographyPrimitivesTypes.Companion_Error_.Create_AwsCryptographicPrimitivesError_(
+			dafny.SeqOfChars([]dafny.Char(err.Error())...)))
+	}
+
+	if paddingMode.Is_PKCS1() {
+		encryptedBytes, err := rsa.DecryptPKCS1v15(rand.Reader, privateKey.(*rsa.PrivateKey), dafny.ToByteArray(cipher_text))
+		if err != nil {
+			return Wrappers.Companion_Result_.Create_Failure_(AwsCryptographyPrimitivesTypes.Companion_Error_.Create_AwsCryptographicPrimitivesError_(
+				dafny.SeqOfChars([]dafny.Char(err.Error())...)))
+		}
+		return Wrappers.Companion_Result_.Create_Success_(dafny.SeqOfBytes(encryptedBytes))
+	}
+
+	mode, err := getNativeHashAlgorithm(paddingMode)
+	if err != nil {
+		return Wrappers.Companion_Result_.Create_Failure_(AwsCryptographyPrimitivesTypes.Companion_Error_.Create_AwsCryptographicPrimitivesError_(
+			dafny.SeqOfChars([]dafny.Char(err.Error())...)))
+	}
+
+	res, err := rsa.DecryptOAEP(crypto.Hash(mode).New(), rand.Reader, privateKey.(*rsa.PrivateKey), dafny.ToByteArray(cipher_text), nil)
+	if err != nil {
+		return Wrappers.Companion_Result_.Create_Failure_(AwsCryptographyPrimitivesTypes.Companion_Error_.Create_AwsCryptographicPrimitivesError_(
+			dafny.SeqOfChars([]dafny.Char(err.Error())...)))
+	}
+	return Wrappers.Companion_Result_.Create_Success_(dafny.SeqOfBytes(res))
+}
+
+func (CompanionStruct_Default___) EncryptExtern(paddingMode AwsCryptographyPrimitivesTypes.RSAPaddingMode,
+		key dafny.Sequence, plainText dafny.Sequence) Wrappers.Result {
+	derPublicKey, rest := pem.Decode(dafny.ToByteArray(key))
```

Comment:
this is technically the key block not der s/derPublicKey/block

Reply:
We do use the der format during the pem encoding - so reverse would be that pem-decode gives us the der formatted key. Also the method`x509.ParsePKIXPublicKey` takes derBytes as input, so the key has to be in der format.

I think the reason pem.Decode says it's a `block` because they don't know the underlying format.